### PR TITLE
Support custom root namespaces

### PR DESCRIPTION
**Pull request by mclark**
Zeitwerk [supports defining root namespaces](https://github.com/fxn/zeitwerk?tab=readme-ov-file#zeitwerkloaderdirs) for autoload root directories. For example, maybe you want to nest all constants defined in `app/services` under the `CompanyName` domain. Under this rule, Zeitwerk would expect a file called `app/services/widget_maker.rb` to define a constant `::CompanyName::WidgetMaker` rather than just `WidgetMaker`.

This pull request allows us to teach `packs` about the overridden namespaces via a new configuration key `namespace_overrides`. It takes the autoload root dirs as keys and the name of the custom root namespace as the values. So for the above example, you'd need to specify the following in `packwerk.yml`:

```
namespace_overrides:
  app/services: "::CompanyName"
```

Two things that might be improved:

* [x] The configuration could store the absolute autoload root directories instead of the relative directories. This would improve performance in the tight loop where we resolve constant definitions from file paths.
* [x] The `HashMap` could maybe use `PathBuf` or something instead of `String`? I don't know, it's been a long time and my Rust is quite... rusty. 🙈 

In addition to the above two things, we could benefit from wrapping a bunch of parameters being passed down from the config in some sort of abstraction. Five parameters is getting a bit much. I figured that is a good opportunity for a follow up PR.

Many many thanks for the help from @mathias and @bk2204 who were immensely helpful when my Rust memory was failing.

This PR goes a long way to getting Packs support for the GitHub monolith, but we still have some work to do! I'd say 90%+ of remaining differences between Packs and Packwerk for the GitHub monolith are resolved with this change.

### CODE REVIEW

File: tests/fixtures/simple_app_with_overrides/packwerk.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+# List of custom associations, if any
+# custom_associations:
+# - "cache_belongs_to"
+
+# Whether or not you want the cache enabled (disabled by default)
+cache: false
+
+namespace_overrides:
+  app/services: "::CustomNamespace"
```
Comment (alexevanczuk) [inline tests/fixtures/simple_app_with_overrides/packwerk.yml:23]: So does this mean that this affects only `/app/services`? Or all of `packs/*/app/services`? Based on the implementation it looks like the former.

I guess we just want to match the zeitwerk implementation here? So I guess that means any relative folder => its custom namespace.

So if you wanted to mark a whole pack as having a custom namespace, would you do:
```
namespace_overrides:
  packs/bar: "::Bar"
```
?

File: tests/fixtures/simple_app_with_overrides/node_modules/file.rb [hunk 2]
```diff
@@ -0,0 +1 @@
+Foo
```
Comment (alexevanczuk) [inline tests/fixtures/simple_app_with_overrides/node_modules/file.rb:1]: nit: for this functionality we probably don't need these files. For test cases I was thinking:
- One integration test with one app that has three cases:
  - A pack that has a complete namespace override
  - A pack with a partial namespace override
  - The root having a partial namespace override
  - maybe something else?

We could also have a unit test for these but I've been liking the integration tests because they allow a lot of flexibility in how the internal structure is represented. Unit tests might be good here if there are various edge cases.

**Review by alexevanczuk — COMMENTED** [review]
This looks good! Notice this is a draft but wanted to leave some comments. The implementation makes sense to me, but wanted to better understand – zeitwerk allows you to specify any arbitrary folder and saw that it has a namespace override right? Does it allow the folder declaration to be a glob?

I'm wondering how you are hoping to use this. Originally I thought maybe it makes sense to have these namespace overrides in `package.yml`, but if this is meant to just be a shim so `packs` can know about something about zeitwerk, it makes sense to emulate the zeitwerk API to some degree. So I'd imagine some sort of ruby tooling that dumps zeitwerk's rules into a `packs` compatible hash for the `packwerk.yml`, is that what you were thinking?

**Comment by mclark** [comment]
👋 Hi @alexevanczuk! Sorry for the long delay responding, I didn't go near the computer for my holidays to give myself a proper break 😄 

> zeitwerk allows you to specify any arbitrary folder and saw that it has a namespace override right? Does it allow the folder declaration to be a glob?

Essentially this change is trying to add support for the optional namespace parameter to [`push_dir`](https://github.com/fxn/zeitwerk/blob/a7201b3d756f341a1d823cf4e7d34ff623e1b17d/lib/zeitwerk/loader/config.rb#L103-L127). That method expects an absolute path but I was planning on going with a relative path for this implementation. There's not reason we couldn't add support for globs though, perhaps as a followup?

> if this is meant to just be a shim so packs can know about something about zeitwerk, it makes sense to emulate the zeitwerk API to some degree.

Yeah you got it.

To be completely transparent, I really like your experimental constant detector/resolver so I would like to move to that. This is just so that the packwerk compatible version of packs works and matches the existing packwerk output for our monolith. Then we'll turn on experimental mode most likely.

> So I'd imagine some sort of ruby tooling that dumps zeitwerk's rules into a packs compatible hash for the packwerk.yml, is that what you were thinking?

That would be very cool, but I'd wait for demand for that before investing any time in it. Like I said, for our purposes at GitHub this is hopefully just a stepping stone.

File: tests/fixtures/simple_app_with_overrides/node_modules/file.rb (hunk 2 above)
Comment (mclark) [reply tests/fixtures/simple_app_with_overrides/node_modules/file.rb:1]: Oh great call out, thanks for reminding me! I probably just would have kept them otherwise. This is the mess you get when you copy/paste file folders 😂

Comment (mclark) [reply tests/fixtures/simple_app_with_overrides/node_modules/file.rb:1]: Do you want these overrides to be specified on a per-pack basis then? I think that is a feasible change for this PR, I just want to be clear before doing the implementation needed.

Comment (alexevanczuk) [reply tests/fixtures/simple_app_with_overrides/node_modules/file.rb:1]: @mclark and I chatted over slack (in the Ruby/Rails modularity slack server, [linked in the README](https://github.com/alexevanczuk/packs#kudos)) and we're thinking putting this into `packwerk.yml`/`packs.yml` makes sense for the first iteration at least.

The main reasoning for this is that we felt like it's valuable to consolidate autoload configuration into one place, tying autoloading more to the application as a whole rather than each package. Either way, I think the decision here is pretty reversible – I could see `packs` allowing users to choose one or the other and having `packs` double-read the information.

**Comment by alexevanczuk** [comment]
Chatted with Matt over slack, adding responses here for posterity.

> I didn't go near the computer for my holidays to give myself a proper break 😄

Glad to hear it!

> That method expects an absolute path but I was planning on going with a relative path for this implementation. There's not reason we couldn't add support for globs though, perhaps as a followup?

All sounds good. Mostly just curious, we can add to it as need dictates!

> To be completely transparent, I really like your experimental constant detector/resolver so I would like to move to that. This is just so that the packwerk compatible version of packs works and matches the existing packwerk output for our monolith. Then we'll turn on experimental mode most likely.

Sounds good. Either way, having `packs` support this `zeitwerk` feature is a huge win in terms of having it be at parity with `packwerk` for more apps, so not sweating the details at the moment. I also think that after `packwerk` adopts `prism` for AST parsing, there might be an option to decouple from `zeitwerk` and use `prism` to parse both references and definitions. It would be really great if both implementations (`packwerk` and `packs`) could only support one parsing definition (the "experimental parser" approach, which seems strictly better than the zeitwerk based approach, although open to hearing and curious about other opinions here).

> That would be very cool, but I'd wait for demand for that before investing any time in it. Like I said, for our purposes at GitHub this is hopefully just a stepping stone.

For sure, one step at a time either way!

**Comment by mclark** [comment]
This is ready for the next stage! 🎉

**Review by alexevanczuk — APPROVED** [review]
This looks great! Great job on this 🎉
